aptos-move/aptos-vm/src/block_executor/mod.rs

```diff
@@ -115,8 +115,96 @@ impl BlockExecutorTransactionOutput for AptosTransactionOutput {
     }
 
     // TODO: get rid of the cloning data-structures in the following APIs.
+    // TODO(BlockSTMv2): convert invariant checks to PanicErrors.
 
-    /// Should never be called after incorporating materialized output, as that consumes vm_output.
+    fn incorporate_materialized_txn_output(
+        &self,
+        aggregator_v1_writes: Vec<(StateKey, WriteOp)>,
+        materialized_resource_write_set: Vec<(StateKey, WriteOp)>,
+        materialized_events: Vec<ContractEvent>,
+    ) -> Result<(), PanicError> {
+        assert!(
+            self.committed_output
+                .set(
+                    self.vm_output
+                        .lock()
+                        .take()
+                        .expect("Output must be set to incorporate materialized data")
+                        .into_transaction_output_with_materialized_write_set(
+                            aggregator_v1_writes,
+                            materialized_resource_write_set,
+                            materialized_events,
+                        )?,
+                )
+                .is_ok(),
+            "Could not combine VMOutput with the materialized resource and event data"
+        );
+        Ok(())
+    }
+
+    fn set_txn_output_for_non_dynamic_change_set(&self) {
+        assert!(
+            self.committed_output
+                .set(
+                    self.vm_output
+                        .lock()
+                        .take()
+                        .expect("Output must be set to incorporate materialized data")
+                        .into_transaction_output()
+                        .expect("We should be able to always convert to transaction output"),
+                )
+                .is_ok(),
+            "Could not combine VMOutput with the materialized resource and event data"
+        );
+    }
+
+    /// TODO(gelash): Consider defensive access pattern to committed_output / vm_output.
+    /// Returns true iff the execution status is Keep(Success).
+    fn is_kept_success(&self) -> Result<bool, PanicError> {
```

Comment:
add a comment to explain that the only reason we cover both committed and vm_output cases is because the first is used by sequential execution - since it's not permissible once we change vm_output to be ExplicitSyncWrapper to have calls concurrent w. materialize (which moves vm_output to committed).